File: 0048-rotate-image/0048-rotate-image.js
/**
 * @param {number[][]} matrix
 * @return {void} Do not return anything, modify matrix in-place instead.
 */
var rotate = function(matrix) {
    const reassign = (matrix, width, corner) => {
        if (width === 1) return;
        const [lb, ub] = [corner, corner + width - 1];
        for (let i = 0; i < width - 1; i++) {
            [
                matrix[lb][lb+i],
                matrix[lb+i][ub],
                matrix[ub][ub-i],
                matrix[ub-i][lb]
            ] = [
                matrix[ub-i][lb],
                matrix[lb][lb+i],
                matrix[lb+i][ub],
                matrix[ub][ub-i]
            ]
        }
    };
    
    for (let w = matrix.length; w > 1; w -= 2) {
        reassign(matrix, w, (matrix.length - w) / 2);
    }
};

File: 0048-rotate-image/0048-rotate-image.go
func rotate(matrix [][]int)  {
  c := 0
  for w:=len(matrix);w>1;w-=2 {
    c = (len(matrix)-w)/2
    reassign(matrix, w, c)
  }
}

func reassign(m [][]int, w int, c int)  {
  if w == 1 {
    return
  }
  lb, ub := c, c+w-1
  for i:=0;i<w-1;i++ {
    m[lb][lb+i], m[lb+i][ub], m[ub][ub-i], m[ub-i][lb] = m[ub-i][lb], m[lb][lb+i], m[lb+i][ub], m[ub][ub-i]
  }
}